File: src/http/messages.ts
import type { Context } from 'hono';
import type { ProviderAdapter, ProviderRequestBody } from '../adapters/provider.js';
import type { CompressionPolicy, PolicyFailure } from '../policy/headers.js';
import type { IrRequest } from '../ir/types.js';
import type { PipelineStats } from '../compression/pipeline.js';
import type { ResponseDecorator } from './upstream.js';
import type { SavingsReporter } from '../telemetry/savings-log.js';
import { parseCompressionPolicy } from '../policy/headers.js';
import { accountFor, applyAccountingHeaders } from '../telemetry/accounting.js';
import { UnknownScorerError } from './unknown-scorer-error.js';
import {
  forwardableRequestHeaders,
  passthroughResponse,
  sendUpstream,
} from './upstream.js';

/**
 * The seam the compression pipeline plugs into. The stage returns its stats so
 * telemetry can label the response without re-walking the request.
 */
export type StageResult = {
  request: IrRequest;
  stats: PipelineStats | null;
};

export type CompressionStage = (
  request: IrRequest,
  policy: CompressionPolicy,
) => StageResult;

export const identityCompressionStage: CompressionStage = (request) => ({
  request,
  stats: null,
});

class MalformedBodyError extends Error {
  constructor(path: string, cause: unknown) {
    super(`parsing ${path} request body as JSON failed`, { cause });
    this.name = 'MalformedBodyError';
  }
}

function policyErrorMessage(failure: PolicyFailure): string {
  return `${failure.header}: ${failure.reason} (received "${failure.value}")`;
}

function unknownScorerMessage(error: UnknownScorerError): string {
  return `X-Caveman-Scorer: unknown scorer "${error.requested}" (available: ${error.available.join(', ')})`;
}

function parseBody(raw: string, path: string): ProviderRequestBody {
  try {
    return JSON.parse(raw) as ProviderRequestBody;
  } catch (cause) {
    throw new MalformedBodyError(path, cause);
  }
}

/** Absent stats mean the request was never compressed, so nothing is reported. */
function accountingDecorator(stats: PipelineStats | null): ResponseDecorator {
  return (headers) => {
    if (stats === null) return;
    applyAccountingHeaders(headers, accountFor(stats));
  };
}

type Route = {
  adapter: ProviderAdapter;
  stage: CompressionStage;
  reporter: SavingsReporter | null;
};

function reject(context: Context, route: Route, message: string): Response {
  return context.json(route.adapter.errorEnvelope(message), 400);
}

async function readBody(context: Context, route: Route): Promise<ProviderRequestBody> {
  return parseBody(await context.req.text(), route.adapter.path);
}

/**
 * The client's query string, forwarded so upstream sees the request it was
 * addressed to. An unparseable URL yields none rather than failing the request.
 */
function incomingSearch(url: string): string {
  try {
    return new URL(url).search;
  } catch {
    return '';
  }
}

async function forwardMessages(context: Context, route: Route): Promise<Response> {
  const policyResult = parseCompressionPolicy(context.req.raw.headers);
  if (!policyResult.ok) {
    return reject(context, route, policyErrorMessage(policyResult));
  }

  let body: ProviderRequestBody;
  try {
    body = await readBody(context, route);
  } catch (error) {
    if (!(error instanceof MalformedBodyError)) throw error;
    return reject(context, route, 'request body is not valid JSON');
  }

  let staged: StageResult;
  try {
    staged = route.stage(route.adapter.toIR(body), policyResult.policy);
  } catch (error) {
    if (!(error instanceof UnknownScorerError)) throw error;
    return reject(context, route, unknownScorerMessage(error));
  }

  const upstream = await sendUpstream({
    adapter: route.adapter,
    headers: forwardableRequestHeaders(context.req.raw.headers),
    body: JSON.stringify(route.adapter.fromIR(staged.request)),
    search: incomingSearch(context.req.raw.url),
    signal: context.req.raw.signal,
  });
  if (staged.stats !== null) {
    route.reporter?.record(staged.stats);
  }
  return passthroughResponse(upstream, accountingDecorator(staged.stats));
}

/**
 * Builds the handler for one provider. The adapter, the pipeline, and the
 * reporter are injected, so the handler holds no knowledge of any provider's
 * wire format and none of where its output goes.
 */
export function createMessagesHandler(
  adapter: ProviderAdapter,
  stage: CompressionStage = identityCompressionStage,
  reporter: SavingsReporter | null = null,
): (context: Context) => Promise<Response> {
  const route: Route = { adapter, stage, reporter };
  return (context) => forwardMessages(context, route);
}

File: internal/server/messages.go
package server

import (
	"io"
	"net/http"

	"github.com/carlelieser/caveman/internal/adapters"
	"github.com/carlelieser/caveman/internal/ir"
	"github.com/carlelieser/caveman/internal/policy"
	"github.com/carlelieser/caveman/internal/telemetry"
)

// StageResult is what one compression pass produced. Stats are nil when the
// request was never compressed, so telemetry can label the response without
// re-walking it.
type StageResult struct {
	Request ir.Request
	Stats   *telemetry.PipelineStats
}

// CompressionStage is the seam the compression pipeline plugs into. The handler
// holds no knowledge of how text is compressed, only of when to ask.
type CompressionStage func(request ir.Request, p policy.Policy) StageResult

// IdentityStage forwards a request unchanged. It is the default, so a server
// built without a pipeline is a transparent proxy rather than a broken one.
func IdentityStage(request ir.Request, _ policy.Policy) StageResult {
	return StageResult{Request: request}
}

type route struct {
	provider adapters.Provider
	stage    CompressionStage
	reporter *telemetry.Reporter
	client   *http.Client
}

func (r route) reject(writer http.ResponseWriter, message string) {
	body := ir.Marshal(r.provider.ErrorEnvelope(message))
	writer.Header().Set("content-type", "application/json")
	writer.WriteHeader(http.StatusBadRequest)
	_, _ = writer.Write(body)
}

// accountingDecorator labels the response with what compression did. Absent
// stats mean the request was never compressed, so nothing is reported.
func accountingDecorator(stats *telemetry.PipelineStats) responseDecorator {
	return func(headers http.Header) {
		if stats == nil {
			return
		}
		telemetry.ApplyAccountingHeaders(headers, telemetry.AccountFor(*stats))
	}
}

// observerFor watches the response for the counts the provider billed.
// Attached even when compression is off, so an uncompressed session gives a
// baseline to compare against.
func observerFor(reporter *telemetry.Reporter) bodyObserver {
	if reporter == nil {
		return nil
	}
	return &usageObserver{observer: telemetry.NewUsageObserver(), reporter: reporter}
}

func (r route) handle(writer http.ResponseWriter, request *http.Request) {
	compressionPolicy, failure := policy.Parse(request.Header)
	if failure != nil {
		r.reject(writer, failure.Error())
		return
	}

	raw, err := io.ReadAll(request.Body)
	if err != nil {
		r.reject(writer, "reading the request body failed")
		return
	}
	parsed, err := ir.Unmarshal(raw)
	if err != nil {
		r.reject(writer, "request body is not valid JSON")
		return
	}
	body, ok := parsed.(*ir.Object)
	if !ok {
		r.reject(writer, "request body is not a JSON object")
		return
	}

	staged := r.stage(r.provider.ToIR(body), compressionPolicy)

	// The client's cancellation reaches upstream, so hanging up here stops the
	// work there rather than leaving it to finish against nobody.
	upstream, err := sendUpstream(request.Context(), r.client, upstreamRequest{
		provider: r.provider,
		headers:  forwardableRequestHeaders(request.Header),
		body:     ir.MarshalString(r.provider.FromIR(staged.Request)),
		search:   incomingSearch(request),
	})
	if err != nil {
		writer.Header().Set("content-type", "application/json")
		writer.WriteHeader(http.StatusBadGateway)
		_, _ = writer.Write(ir.Marshal(r.provider.ErrorEnvelope("upstream request failed")))
		return
	}
	defer upstream.Body.Close()

	if staged.Stats != nil && r.reporter != nil {
		r.reporter.Record(*staged.Stats)
	}
	passthroughResponse(writer, upstream, accountingDecorator(staged.Stats), observerFor(r.reporter))
}

// incomingSearch is the client's query string, forwarded so upstream sees the
// request it was addressed to.
func incomingSearch(request *http.Request) string {
	if request.URL.RawQuery == "" {
		return ""
	}
	return "?" + request.URL.RawQuery
}
